File: OpenDentBusiness/TableTypes/HL7DefSegment.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml.Serialization;

namespace OpenDentBusiness {
	///<summary>multiple segments per message</summary>
	[Serializable]
	public class HL7DefSegment:TableBase {
		///<summary>Primary key.</summary>
		[CrudColumn(IsPriKey=true)]
		public long HL7DefSegmentNum;
		///<summary>FK to HL7DefMessage.HL7DefMessageNum</summary>
		public long HL7DefMessageNum;
		///<summary>Since we don't enforce or automate, it can be 1-based or 0-based.  For outgoing, this affects the message structure.  For incoming, this is just for convenience and organization in the HL7 Def windows.</summary>
		public int ItemOrder;
		///<summary>For example, a DFT can have multiple FT1 segments.</summary>
		public bool CanRepeat;
		///<summary>An incoming message may or may not contain this segment. Not used for outgoing.</summary>
		public bool IsOptional;
		///<summary>Stored in db as string, but used in OD as enum SegmentNameHL7. Example: PID.</summary>
		[CrudColumn(SpecialType=CrudSpecialColType.EnumAsString)]
		public SegmentNameHL7 SegmentName;
		///<summary>.</summary>
		[CrudColumn(SpecialType=CrudSpecialColType.TextIsClob)]
		public string Note;

		///<Summary>List of segments associated with this hierarchical definition.  Use items in this list to get to items lower in the hierarchy.</Summary>
		[CrudColumn(IsNotDbColumn=true)]
		[XmlIgnore]
		public List<HL7DefField> hl7DefFields;

		///<summary></summary>
		public HL7DefSegment Clone() {
			return (HL7DefSegment)this.MemberwiseClone();
		}

		public void AddField(int ordinalPos,string tableId,DataTypeHL7 dataType,string fieldName,string fixedText) {
			if(hl7DefFields==null){
				hl7DefFields=new List<HL7DefField>();
			}
			HL7DefField field=new HL7DefField();
			field.OrdinalPos=ordinalPos;
			field.DataType=dataType;
			field.TableId=tableId;
			field.FieldName=fieldName;
			field.FixedText=fixedText;
			this.hl7DefFields.Add(field);
		}

		public void AddField(int ordinalPos,DataTypeHL7 dataType,string fieldName) {
			AddField(ordinalPos,"",dataType,fieldName,"");
		}

		/// <summary></summary>
		public void AddFieldFixed(int ordinalPos,DataTypeHL7 dataType,string fixedText) {
			AddField(ordinalPos,"",dataType,"",fixedText);
		}

	}

	///<summary>The items in this enumeration can be freely rearranged without damaging the database.  But can't change spelling or remove existing item.</summary>
	public enum SegmentNameHL7 {
		///<summary>Db Resource Appointment Information</summary>
		AIG,
		///<summary>Location Resource Appointment Information</summary>
		AIL,
		///<summary>Personnel Resource Appointment Information</summary>
		AIP,
		///<summary>Diagnosis Information</summary>
		DG1,
		///<summary>Event Type</summary>
		EVN,
		///<summary>Financial Transaction Information</summary>
		FT1,
		///<summary>Guarantor Information</summary>
		GT1,
		///<summary>Insurance Information</summary>
		IN1,
		///<summary>Message Header</summary>
		MSH,
		///<summary>Observations Request</summary>
		OBR,
		///<summary>Observation Related to OBR</summary>
		OBX,
		///<summary>Common Order.  Used in outgoing vaccinations VXUs as well as incoming lab result ORUs.</summary>
		ORC,
		///<summary>Patient Identification</summary>
		PID,
		///<summary>Patient additional demographics</summary>
		PD1,
		///<summary>Patient Visit</summary>
		PV1,
		///<summary>Pharmacy Administration Segment</summary>
		RXA,
		///<summary>Scheduling Activity Information</summary>
		SCH,
		///<summary>This can happen for unsupported segments.</summary>
		Unknown,
		///<summary>We use for PDF Data</summary>
		ZX1,
	}
}

File: OpenDentalWeb/OpenDental/src/com/opendental/odweb/client/tabletypes/HL7DefSegment.java
package com.opendental.odweb.client.tabletypes;

import com.google.gwt.xml.client.Document;
import com.opendental.odweb.client.remoting.Serializing;
import java.util.ArrayList;

//DO NOT MAKE CHANGES TO THIS FILE.  THEY WILL GET OVERWRITTEN BY THE CRUD.
public class HL7DefSegment {
		/** Primary key. */
		public int HL7DefSegmentNum;
		/** FK to HL7DefMessage.HL7DefMessageNum */
		public int HL7DefMessageNum;
		/** Since we don't enforce or automate, it can be 1-based or 0-based.  For outgoing, this affects the message structure.  For incoming, this is just for convenience and organization in the HL7 Def windows. */
		public int ItemOrder;
		/** For example, a DFT can have multiple FT1 segments.  This turns out to be a completely useless field, since we already know which ones can repeat. */
		public boolean CanRepeat;
		/** If this is false, and an incoming message is missing this segment, then it gets logged as an error/failure.  If this is true, then it will gracefully skip a missing incoming segment.  Not used for outgoing. */
		public boolean IsOptional;
		/** Stored in db as string, but used in OD as enum SegmentNameHL7. Example: PID. */
		public SegmentNameHL7 SegmentName;
		/** . */
		public String Note;
		/**  */
		public ArrayList<HL7DefField> hl7DefFields;

		/** Deep copy of object. */
		public HL7DefSegment deepCopy() {
			HL7DefSegment hl7defsegment=new HL7DefSegment();
			hl7defsegment.HL7DefSegmentNum=this.HL7DefSegmentNum;
			hl7defsegment.HL7DefMessageNum=this.HL7DefMessageNum;
			hl7defsegment.ItemOrder=this.ItemOrder;
			hl7defsegment.CanRepeat=this.CanRepeat;
			hl7defsegment.IsOptional=this.IsOptional;
			hl7defsegment.SegmentName=this.SegmentName;
			hl7defsegment.Note=this.Note;
			hl7defsegment.hl7DefFields=this.hl7DefFields;
			return hl7defsegment;
		}

		/** Serialize the object into XML. */
		public String serialize() {
			StringBuilder sb=new StringBuilder();
			sb.append("<HL7DefSegment>");
			sb.append("<HL7DefSegmentNum>").append(HL7DefSegmentNum).append("</HL7DefSegmentNum>");
			sb.append("<HL7DefMessageNum>").append(HL7DefMessageNum).append("</HL7DefMessageNum>");
			sb.append("<ItemOrder>").append(ItemOrder).append("</ItemOrder>");
			sb.append("<CanRepeat>").append((CanRepeat)?1:0).append("</CanRepeat>");
			sb.append("<IsOptional>").append((IsOptional)?1:0).append("</IsOptional>");
			sb.append("<SegmentName>").append(SegmentName.ordinal()).append("</SegmentName>");
			sb.append("<Note>").append(Serializing.escapeForXml(Note)).append("</Note>");
			sb.append("</HL7DefSegment>");
			return sb.toString();
		}

		/** Sets all the variables on this object based on the values in the XML document.  Variables that are not in the XML document will be null or their default values.
		 * @param doc A parsed XML document.  Must be valid XML.  Does not need to contain a node for every variable on this object.
		 * @throws Exception DeserializeFromXml is entirely encased in a try catch and will throw exceptions if anything goes wrong. */
		public void deserialize(Document doc) throws Exception {
			try {
				if(Serializing.getXmlNodeValue(doc,"HL7DefSegmentNum")!=null) {
					HL7DefSegmentNum=Integer.valueOf(Serializing.getXmlNodeValue(doc,"HL7DefSegmentNum"));
				}
				if(Serializing.getXmlNodeValue(doc,"HL7DefMessageNum")!=null) {
					HL7DefMessageNum=Integer.valueOf(Serializing.getXmlNodeValue(doc,"HL7DefMessageNum"));
				}
				if(Serializing.getXmlNodeValue(doc,"ItemOrder")!=null) {
					ItemOrder=Integer.valueOf(Serializing.getXmlNodeValue(doc,"ItemOrder"));
				}
				if(Serializing.getXmlNodeValue(doc,"CanRepeat")!=null) {
					CanRepeat=(Serializing.getXmlNodeValue(doc,"CanRepeat")=="0")?false:true;
				}
				if(Serializing.getXmlNodeValue(doc,"IsOptional")!=null) {
					IsOptional=(Serializing.getXmlNodeValue(doc,"IsOptional")=="0")?false:true;
				}
				if(Serializing.getXmlNodeValue(doc,"SegmentName")!=null) {
					SegmentName=SegmentNameHL7.valueOf(Serializing.getXmlNodeValue(doc,"SegmentName"));
				}
				if(Serializing.getXmlNodeValue(doc,"Note")!=null) {
					Note=Serializing.getXmlNodeValue(doc,"Note");
				}
			}
			catch(Exception e) {
				throw new Exception("Error deserializing HL7DefSegment: "+e.getMessage());
			}
		}

		/** The items in this enumeration can be freely rearranged without damaging the database.  But can't change spelling or remove existing item. */
		public enum SegmentNameHL7 {
			/** Db Resource Appointment Information */
			AIG,
			/** Location Resource Appointment Information */
			AIL,
			/** Personnel Resource Appointment Information */
			AIP,
			/** Diagnosis Information */
			DG1,
			/** Event Type */
			EVN,
			/** Financial Transaction Information */
			FT1,
			/** Guarantor Information */
			GT1,
			/** Insurance Information */
			IN1,
			/** Message Header */
			MSH,
			/** Observations Request */
			OBR,
			/** Observation Related to OBR */
			OBX,
			/** Common Order.  Used in outgoing vaccinations VXUs as well as incoming lab result ORUs. */
			ORC,
			/** Patient Identification */
			PID,
			/** Patient additional demographics */
			PD1,
			/** Patient Visit */
			PV1,
			/** Pharmacy Administration Segment */
			RXA,
			/** Scheduling Activity Information */
			SCH,
			/** This can happen for unsupported segments. */
			Unknown,
			/** We use for PDF Data */
			ZX1
		}


}
